### PR TITLE
Updating `CONTRIBUTING.md` and `STYLE_GUIDE.md`

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
- Add links to typing documentation in the style guide
- Added a note to document class `__init__` args under the class-level docstring
- Added in some notes about code structuring
- Fixed links (using relative links for a few items in the repo; switched the link in CONTRIBUTING.md to point to our style guide rather than the google one)
- Added a note about that callers, rather than calles, should check for argument values that result in no-ops

### CODE REVIEW

File: STYLE_GUIDE.md [hunk 1]
```diff
@@ -36,12 +36,32 @@ pre-commit run isort --all-files  # for isort
 The configuration is stored in [pyproject.toml](pyproject.toml).
 
 
+### 2.4. Code Structure
+
+As a general rule of thumb,
+
+- Don't: use classes to make code look more like an outline
```
Comment (hanlint) [inline STYLE_GUIDE.md:43]: what does this mean?

File: STYLE_GUIDE.md [hunk 2]
```diff
@@ -274,6 +294,7 @@ Composer uses [Google Style Docstrings](https://sphinxcontrib-napoleon.readthedo
 
 The following guidelines apply to documentation.
 1.  Each function that needs a docstring must have its input arguments and return statement (if not None) annotated.
```
Comment (eracah) [inline STYLE_GUIDE.md:296]: And `Raises`?

**Review by eracah — APPROVED** [review]

File: STYLE_GUIDE.md (hunk 1 above)
Comment (ravi-mosaicml) [reply STYLE_GUIDE.md:43]: Lol, good point. I think it had something to do with the last bullet, but yea let me remove it. Copied it from https://www.notion.so/Codebase-74ccd984a16c4095a368806225a555b4#23cafb24f4344751a700cd875dc095e9

File: STYLE_GUIDE.md (hunk 2 above)
Comment (ravi-mosaicml) [reply STYLE_GUIDE.md]: Added a comment that all custom exceptions should be documented.

File: STYLE_GUIDE.md (hunk 1 above)
Comment (ravi-mosaicml) [reply STYLE_GUIDE.md]: Fixed

**Pull request merged**